extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -85,52 +88,104 @@ private CosmosDBInputAttribute GetBindingDataContent(ModelBindingData bindingDat
 
         private async Task<object> CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute)
         {
-            MethodInfo createPOCOMethod;
+            var container = CreateCosmosClient<Container>(cosmosAttribute) as Container;
 
-            if (targetType.GenericTypeArguments.Any())
+            if (container is null)
             {
-                targetType = targetType.GenericTypeArguments.FirstOrDefault();
+                throw new InvalidOperationException($"Unable to create Cosmos container client for '{cosmosAttribute.ContainerName}'.");
+            }
 
-                createPOCOMethod = GetType()
-                                    .GetMethod(nameof(CreatePOCOCollectionAsync), BindingFlags.Instance | BindingFlags.NonPublic)
-                                    .MakeGenericMethod(targetType);
+            if (targetType.TryGetCollectionElementType(out Type? elementType))
+            {
+                if (elementType is null)
+                {
+                    throw new ArgumentNullException(nameof(elementType));
+                }
+
+                if (targetType.IsConcreteType())
+                {
+                    return await CreatePocoListConcreteAsync(container, cosmosAttribute, elementType, targetType);
```

Comment:
Will this work for arrays? I think `IsConcereteType()` will return `true` for that

Reply:
I will do some E2E testing, it's been a while since all the last couple changes 😞

Reply:
Is it okay to do this `if (targetType.IsConcreteType() && !targetType.IsArray)`? And let arrays fall into the other branch? 

This is mostly working right now, the biggest issue is that all of the streams once deserialized have null properties :/ So trying to figure out why that might be happening

Reply:
Ended up figuring this out and currently in deserialization hell. Turns out the content of "stream" is a json object with an array of documents so it's not so easy to just deserialize to a list/target type. E.g.:

```json
{{
  "_rid": "ltwyAJGNSgs=",
  "Documents": [
    {
      "id": "1",
      "description": "take out the rubbish",
      "_rid": "ltwyAJGNSgsBAAAAAAAAAA==",
      "_self": "dbs/ltwyAA==/colls/ltwyAJGNSgs=/docs/ltwyAJGNSgsBAAAAAAAAAA==/",
      "_etag": "\"0f00f247-0000-0700-0000-63bc7ae60000\"",
      "_attachments": "attachments/",
      "_ts": 1673296614
    }
  ],
  "_count": 1
}}
```

So will need to extract Documents from the stream first, then deserialize that to a list. And will also need to do AddRange and not Add (like I was doing before this refactor using `GetItemQueryIterator` i.e.

```cs
// pre refactor code
private async Task<IList<T>> ExtractCosmosDocumentsAsync<T>(FeedIterator<T> iterator)
        {
            var documentList = new List<T>();
            while (iterator.HasMoreResults)
            {
                FeedResponse<T> response = await iterator.ReadNextAsync();
                documentList.AddRange(response.Resource);
            }
            return documentList;
        }
```

Reply:
Resolved this by providing a type to extract the `Document` property.